Add 2 New Use Case Guides - Pre-Seeding database and WireMock (#20968)
docker/docs: content/guides/use-case/wiremock.md

### Prerequisites

The following prerequisites are required to follow along with this how-to guide:

* [Docker Desktop](https://www.docker.com/products/docker-desktop/)

### Launching WireMock

Launch a quick demo of WireMock by using the following steps:

 1. Clone the [GitHub repository](https://github.com/dockersamples/wiremock-node-docker) locally.

    ```console
    $ git clone https://github.com/dockersamples/wiremock-node-docker
    ```

 2. Navigate to the `wiremock-endpoint` directory

    ```console
    $ cd wiremock-node-docker/
    ```

    WireMock acts as the mock API that your backend will communicate with to retrieve data. The mock API responses have already been created for you in the mappings directory. 

 3. Start the Compose stack by running the following command at the root of the cloned project directory

    ```console
    $ docker compose up -d
    ```

    After a moment, the application will be up and running. 

    ![Diagram showing the WireMock container running on Docker Desktop ](./images/wiremock-using-docker.webp)

>     You can check the logs by selecting the **WireMock** container:

Reviewer: ```suggestion
    You can check the logs by selecting the `wiremock-node-docker` container:
```

    ![Diagram showing the logs of WireMock container running on Docker Desktop ](./images/wiremock-logs-docker-desktop.webp)

 4. Test the Mock API.

    ```console
    $ curl http://localhost:8080/api/v1/getWeather\?city\=Bengaluru
    ```
 

>     It will return the following response:

Reviewer: ```suggestion
    It will return the following canned response with mock data:
```

    ```plaintext
    {"city":"Bengaluru","temperature":27.1,"conditions":"Mostly cloudy","forecasts":[{"date":"2024-09-02T07:00:00+05:30","temperature":83,"conditions":"Partly sunny w/ t-storms"},{"date":"2024-09-03T07:00:00+05:30","temperature":83,"conditions":"Thunderstorms"},{"date":"2024-09-04T07:00:00+05:30","temperature":83,"conditions":"Intermittent clouds"},{"date":"2024-09-05T07:00:00+05:30","temperature":82,"conditions":"Dreary"},{"date":"2024-09-06T07:00:00+05:30","temperature":82,"conditions":"Dreary"}]}
    ```

>     You get this response due to the mapping file in the `wiremock-endpoint/mappings/getWeather/getWeatherBengaluru.json` file. 

Reviewer: ```suggestion
    With WireMock, you define canned responses using mapping files.
    For this request, the mock data is defined in the JSON file at
    `wiremock-endpoint/mappings/getWeather/getWeatherBengaluru.json`.
    
    For more information about stubbing canned responses, refer to the
    [WireMock documentation](https://wiremock.org/docs/stubbing/).
```

## Using WireMock in development

>    Now that you have tried WireMock, let’s use it in development and testing. In this example, you will use a sample application that has a Node.js backend. This app stack has the following configuration:

Reviewer: This paragraph and the content that follows should not be indented

```suggestion
Now that you have tried WireMock, let’s use it in development and testing. In this example, you will use a sample application that has a Node.js backend. This app stack has the following configuration:
```

  * **Local Development Environment**: The context in which the Node.js backend and WireMock are running.
  * **Node.js Backend**: Represents the backend application that handles HTTP requests.
  * **External AccuWeather API**: The real API from which live weather data is fetched.

>   * **WireMock**: The mock server that simulates the API responses during testing. It runs as a Docker container.

Reviewer: Do not bold these, and do not use title case

Author: Removed.

  ![Diagram showing the architecture of WireMock in development ](./images/wiremock-arch.webp)

  * The API Request is sent from the Node.js backend to the External AccuWeather API, which then returns the API Response.
  * Alternatively, when using WireMock, a Mocked API Request is sent to the WireMock server, and a Mocked API Response is returned.

  In this guide, you’ll learn how to:

  * Set up a non-containerized Node app to route requests to the containerized WireMock 
  * Set up a non-containerized Node app to fetch data from the external AccuWeather API.

>   The application can be accessed at [dockersamples/wiremock-node-docker](https://github.com/dockersamples/wiremock-node-docker).

Reviewer: We're already several sections into the guide at this point. This feels like it would belong in an introduction.

Author: I have fixed it by adding the list of guides in the initial para.

> ## Set up a non-containerized Node app to route requests to a containerized WireMock

Reviewer: Is this meant to be a third-level section?

Author: Removed the statement as it is not required.

> Let’s set up a non-containerized Node app to route requests to the WireMock container instead of the external AccuWeather API. Assuming that Node is already installed on your system, follow the steps:

Reviewer: Node and npm are not listed as prerequisites for this guide.

Author: Added a separate instruction for NPM and Node and added link too.

   ```console
   npm install 
   npm run start
   ```
 
   You should see the following output:

    ```plaintext
    > express-api-starter@1.2.0 start
    > node src/index.js

    API_ENDPOINT_BASE: http://localhost:8080
    ..
    Listening: http://localhost:5000
    ```

   The output indicates that your Node application has successfully started. It highlights the project name and the `index.js` file as the application's entry point. The `index.js` file initiates the `getWeather.js` file located within the `src/api/` directory by utilizing the `require` function to import the `api` module and subsequently invoking the `getWeather` function.

   The `getWeather.js` file plays a pivotal role within your Node.js application. The code commences by incorporating the `dotenv` module, which instructs the application to load environment variables from a `.env` file. Depending on the value of the `API_ENDPOINT_BASE` variable, the application intelligently directs requests either to the WireMock server (if set to `http://localhost:8080`) or the AccuWeather API. In this scenario, the request is being routed to the WireMock server.

>     This code snippet ensures that the `ACCUWEATHER_API_KEY` is only required when the application is not using WireMock, preventing errors and improving efficiency.

Reviewer: * This feels very GenAI
* Why are we describing the program as part of step 3? I think this should come before. The step list should demonstrate the "how-to", but this and the code excerpt is an explanation.
* We should clarify that the compose service launched in the previous section is a dependency here.

Author: I have moved this to Step 3. Highlighted the necessary points and kept it in simple words

Author: Also, added a prereq step for WireMock to be up and running .

   ```

   You should see the following output:

   ```plaintext
   {"city":"Bengaluru","temperature":27.1,"conditions":"Mostly cloudy","forecasts":[{"date":"2024-09-02T07:00:00+05:30","temperature":83,"conditions":"Partly sunny w/ t-storms"},{"date":"2024-09-03T07:00:00+05:30","temperature":83,"conditions":"Thunderstorms"},{"date":"2024-09-04T07:00:00+05:30","temperature":83,"conditions":"Intermittent clouds"},{"date":"2024-09-05T07:00:00+05:30","temperature":82,"conditions":"Dreary"},{"date":"2024-09-06T07:00:00+05:30","temperature":82,"conditions":"Dreary"}]}%
   ```

   This indicates that your Node.js application is now successfully routing requests to the WireMock container and receiving the mocked responses

   You might have noticed that you’re trying to use `http://localhost:5000` as the URL instead of port `8080`. This is because your Node.js application is running on port `5000`, and it's routing requests to the WireMock container that's listening on port `8080`.

   > [!TIP]
   > Before you proceed to the next step, ensure that you stop the node application service.

## Set up a non-containerized Node app to fetch data from the external AccuWeather API 

>    To enhance your Node.js application with real-time weather data, you can seamlessly integrate the AccuWeather API. This section of the guide will walk you through the steps involved in setting up a non-containerized Node.js application and fetching weather information directly from the AccuWeather API.

Reviewer: This almost seems like it's a logical progression of the application, whereas I would suggest framing it as two different modes of operation:

- Use mock data in local development
- Use a live API in production

I think what we might want to show is how you can configure your project to select the appropriate mode for what you're doing. For example, using npm scripts:

- `npm run dev` which runs the app in `NODE_ENV=development`, and uses the mock API via compose
- `npm run start` which runs `NODE_ENV=production` and thereby the production config of the app, toward the live API

Maybe there is a more idiomatic way to do it, but you get my point.

Author: Hello @dvdksn I tried modifying the initial para with the following:

``` 
In this guide, you'll learn how to:

- Use Docker to launch up a WireMock container.
- Use mock data in the local development without relying on an external API
- Use a Live API in production to fetch real-time weather data from AccuWeather.
```

Author: I was going through your 2nd point where you suggested using appropriate modes. 
As we're focusing on container-supported development workflow and not using Compose for the apps, I am wondering how to pass the NODE_ENV=development in that case.

Reviewer: in `package.json`:

```
  "scripts": {
    "start": "NODE_ENV=production node src/index.js",
    "dev": "NODE_ENV=development node src/index.js"
  },
```

Author: @dvdksn LMK if there is any action item from my end. Does it look good?

Author: I updated https://github.com/dockersamples/wiremock-node-docker/blob/main/accuweather-api/package.json with the entries.

1. Create an AccuWeather API Key

   Sign up for a free AccuWeather developer account at[https://developer.accuweather.com/](https://developer.accuweather.com/). Within your account, create a new app by selecting `MY APPS` on the top navigation menu to get your unique API key.

   ![Diagram showing the AccuWeather Dashboard](images/wiremock-accuweatherapi.webp)

    [AccuWeather API](https://developer.accuweather.com/) is a web API that provides real-time weather data and forecasts. Developers can use this API to integrate weather information into their applications, websites, or other projects.

2. Change directory to `accuweather-api`

   ```console
   $ cd accuweather-api
   ```

3. Set your AccuWeather API key using the `.env` file:

   > [!TIP]
   >  To prevent conflicts, ensure that any existing environment variables named `API_ENDPOINT_BASE` or `ACCUWEATHER_API_KEY` are removed before modifying the `.env` file.

   Run the following command on your terminal:

   ```console
   unset API_ENDPOINT_BASE
   unset ACCUWEATHER_API_KEY
   ```

   It’s time to set the environment variables in the `.env` file:

> 4. Install the dependencies

Reviewer: ?

Author: Added missing instructions.

    [WireMock documentation](https://wiremock.org/docs/stubbing/).

## Using WireMock in development

Now that you have tried WireMock, let’s use it in development and testing. In this example, you will use a sample application that has a Node.js backend. This app stack has the following configuration:

  - Local Development Environment: The context in which the Node.js backend and WireMock are running.
  - Node.js Backend: Represents the backend application that handles HTTP requests.
  - External AccuWeather API: The real API from which live weather data is fetched.
  - WireMock: The mock server that simulates the API responses during testing. It runs as a Docker container.

  ![Diagram showing the architecture of WireMock in development ](./images/wiremock-arch.webp)

  - In development, the Node.js backend sends a request to WireMock instead of the actual AccuWeather API.
  - In production, it connects directly to the live AccuWeather API for real data.

## Use mock data in local development

Let’s set up a Node app to send requests to the WireMock container instead of the actual AccuWeather API.

### Prerequisite

> - Install [Node and NPM](https://docs.npmjs.com/downloading-and-installing-node-js-and-npm)

Reviewer: ```suggestion
- Install [Node.js and npm](https://docs.npmjs.com/downloading-and-installing-node-js-and-npm)
```

   API_ENDPOINT_BASE=http://localhost:8080
   ```

   This will tell your Node.js application to use the WireMock server for API calls. 

3. Examine the Application Entry Point

   - The main file for the application is `index.js`, located in the `accuweather-api/src/api` directory.
   - This file starts the `getWeather.js` module, which is essential for your Node.js application. It uses the `dotenv` package to load environment variables from the`.env` file.
   - Based on the value of `API_ENDPOINT_BASE`, the application routes requests either to the WireMock server (`http://localhost:8080`) or the AccuWeather API. In this setup, it uses the WireMock server.
   - The code ensures that the `ACCUWEATHER_API_KEY` is required only if the application is not using WireMock, enhancing efficiency and avoiding errors.

    ```javascript
    require("dotenv").config();

    const express = require("express");
    const axios = require("axios");

    const router = express.Router();
    const API_ENDPOINT_BASE = process.env.API_ENDPOINT_BASE;
    const API_KEY = process.env.ACCUWEATHER_API_KEY;

    console.log('API_ENDPOINT_BASE:', API_ENDPOINT_BASE);  // Log after it's defined

>     console.log('ACCUWEATHER_API_KEY:', API_KEY);          // Log after it's defined

Reviewer: We shouldn't recommend logging the API key

Author: @dvdksn Most of the application code is purposely kept simple to keep the focus on the guide's content and should not be considered production-ready. We also captured it in the form of statement under GitHub repo https://github.com/dockersamples/wiremock-node-docker?tab=readme-ov-file#project-overview 

But said that,  it's a best practice. Logging sensitive information like API keys can be a security risk. It’s important to avoid exposing such details in logs, especially in production environments. I can add that point as a tip. What do you think?

Author: Anyway, I went ahead and modified it to:

```
console.log('ACCUWEATHER_API_KEY is set:', !!API_KEY); //  i Log boolean instead of actual key
```

Now it doesn't log the keys:

```
API_ENDPOINT_BASE: http://dataservice.accuweather.com/
ACCUWEATHER_API_KEY is set: true
Listening: http://localhost:5000/
```

## Using WireMock in development

Now that you have tried WireMock, let’s use it in development and testing. In this example, you will use a sample application that has a Node.js backend. This app stack has the following configuration:

  - Local Development Environment: The context in which the Node.js backend and WireMock are running.
  - Node.js Backend: Represents the backend application that handles HTTP requests.
  - External AccuWeather API: The real API from which live weather data is fetched.
  - WireMock: The mock server that simulates the API responses during testing. It runs as a Docker container.

  ![Diagram showing the architecture of WireMock in development ](./images/wiremock-arch.webp)

  - In development, the Node.js backend sends a request to WireMock instead of the actual AccuWeather API.
  - In production, it connects directly to the live AccuWeather API for real data.

## Use mock data in local development

Let’s set up a Node app to send requests to the WireMock container instead of the actual AccuWeather API.

### Prerequisite

- Install [Node.js and npm](https://docs.npmjs.com/downloading-and-installing-node-js-and-npm)

> - Ensure that WireMock container is up and running. (as setup earlier).

Reviewer: ```suggestion
- Ensure that WireMock container is up and running (see [Launching Wiremock](#launching-wiremock)
```

   API_ENDPOINT_BASE=http://localhost:8080
   ```

   This will tell your Node.js application to use the WireMock server for API calls. 

3. Examine the Application Entry Point

   - The main file for the application is `index.js`, located in the `accuweather-api/src/api` directory.
   - This file starts the `getWeather.js` module, which is essential for your Node.js application. It uses the `dotenv` package to load environment variables from the`.env` file.
   - Based on the value of `API_ENDPOINT_BASE`, the application routes requests either to the WireMock server (`http://localhost:8080`) or the AccuWeather API. In this setup, it uses the WireMock server.
   - The code ensures that the `ACCUWEATHER_API_KEY` is required only if the application is not using WireMock, enhancing efficiency and avoiding errors.

    ```javascript
    require("dotenv").config();

    const express = require("express");
    const axios = require("axios");

    const router = express.Router();
    const API_ENDPOINT_BASE = process.env.API_ENDPOINT_BASE;
    const API_KEY = process.env.ACCUWEATHER_API_KEY;

    console.log('API_ENDPOINT_BASE:', API_ENDPOINT_BASE);  // Log after it's defined

>     console.log('ACCUWEATHER_API_KEY:', API_KEY);          // Log after it's defined

Reviewer: ```suggestion
    console.log('ACCUWEATHER_API_KEY is set:', !!API_KEY); // Log boolean instead of actual key
```

    if (!API_ENDPOINT_BASE) {
      throw new Error("API_ENDPOINT_BASE is not defined in environment variables");
    }

    // Only check for API key if not using WireMock
    if (API_ENDPOINT_BASE !== 'http://localhost:8080' && !API_KEY) {
      throw new Error("ACCUWEATHER_API_KEY is not defined in environment variables");
    }
    // Function to fetch the location key for the city
    async function fetchLocationKey(townName) {
      const { data: locationData } = await
    axios.get(`${API_ENDPOINT_BASE}/locations/v1/cities/search`, {
        params: { q: townName, details: false, apikey: API_KEY },
      });
      return locationData[0]?.Key;
    }
    ```  

> 
>     > [!TIP]
>     > For the production environment, it's recommended not to expose API keys details in logs. Instead, you can log a placeholder message to indicate whether the API key is loaded:
> 
>       ```javascript
>       const API_KEY = process.env.ACCUWEATHER_API_KEY;
> 
>       console.log('API_ENDPOINT_BASE:', API_ENDPOINT_BASE);  // Log after it's defined
>       console.log('ACCUWEATHER_API_KEY is set:', !!API_KEY); // Log a boolean value instead of the actual key
> 
>       if (!API_ENDPOINT_BASE) {
>         throw new Error("API_ENDPOINT_BASE is not defined in environment variables");
>       }
> 
>       // Only check for API key if not using WireMock
>       if (API_ENDPOINT_BASE !== 'http://localhost:8080' && !API_KEY) {
>         throw new Error("ACCUWEATHER_API_KEY is not defined in environment variables");
>       }
>       ```
> 
>     > This approach ensures that you only log whether the key is set without revealing its value.

Reviewer: I think we can remove this since the sample code uses this format out-of-the-box

```suggestion
```

4. Start the Node server

   Before you start the Node server, ensure that you have already installed the node packages listed in the package.json file by running `npm install`. 

   ```console
   npm install 

>    npm run start

Reviewer: This will fail on macOS because apple is already running a service on port 5000. Let's change the default port to `5001` in `accuweather-api/src/index.js`

Author: Makese sense. I went ahead and changed the port to 5001.

   ```
 
   You should see the following output:

    ```plaintext
    > express-api-starter@1.2.0 start
    > node src/index.js

    API_ENDPOINT_BASE: http://localhost:8080
    ..

>     Listening: http://localhost:5000

Reviewer: ```suggestion
    Listening: http://localhost:5001
```

    ```

   The output indicates that your Node application has successfully started. 
   Keep this terminal window open. 

5. Test the Mocked API

   Open a new terminal window and run the following command to test the mocked API:

   ```console

>    $ curl "http://localhost:5000/api/v1/getWeather?city=Bengaluru"

Reviewer: ```suggestion
   $ curl "http://localhost:5001/api/v1/getWeather?city=Bengaluru"
```

   ```

   You should see the following output:

   ```plaintext
   {"city":"Bengaluru","temperature":27.1,"conditions":"Mostly cloudy","forecasts":[{"date":"2024-09-02T07:00:00+05:30","temperature":83,"conditions":"Partly sunny w/ t-storms"},{"date":"2024-09-03T07:00:00+05:30","temperature":83,"conditions":"Thunderstorms"},{"date":"2024-09-04T07:00:00+05:30","temperature":83,"conditions":"Intermittent clouds"},{"date":"2024-09-05T07:00:00+05:30","temperature":82,"conditions":"Dreary"},{"date":"2024-09-06T07:00:00+05:30","temperature":82,"conditions":"Dreary"}]}%
   ```

   This indicates that your Node.js application is now successfully routing requests to the WireMock container and receiving the mocked responses

>    You might have noticed that you’re trying to use `http://localhost:5000` as the URL instead of port `8080`. This is because your Node.js application is running on port `5000`, and it's routing requests to the WireMock container that's listening on port `8080`.

Reviewer: ```suggestion
   You might have noticed that you’re trying to use `http://localhost:5001` as the URL instead of port `8080`. This is because your Node.js application is running on port `5001`, and it's routing requests to the WireMock container that's listening on port `8080`.
```

   Run the following command on your terminal:

   ```console
   unset API_ENDPOINT_BASE
   unset ACCUWEATHER_API_KEY
   ```

   It’s time to set the environment variables in the `.env` file:

   ```plaintext
   ACCUWEATHER_API_KEY=XXXXXX
   API_ENDPOINT_BASE=http://dataservice.accuweather.com
   ``` 

   Make sure to populate `ACCUWEATHER_API_KEY` with the correct value.

4. Install the dependencies

   Run the following command to install the required packages:

   ```console
   $ npm install
   ```

   This will install all the packages listed in your `package.json` file. These packages are essential for the project to function correctly.

   If you encounter any warnings related to deprecated packages, you can ignore them for now for this demonstration.

5. Assuming that you don’t have a pre-existing Node server running on your system, go ahead and start the Node server by running the following command:

   ```console
   $ npm run start
   ```
  
   You should see the following output:

   ```plaintext
   > express-api-starter@1.2.0 start
   > node src/index.js

   API_ENDPOINT_BASE: http://dataservice.accuweather.com
   ACCUWEATHER_API_KEY is set: true 

>    Listening: http://localhost:5000

Reviewer: ```suggestion
   Listening: http://localhost:5001
```
